### PR TITLE
RegexFind

### PR DESCRIPTION
**Pull request by zwim**
In addition to https://github.com/koreader/crengine/pull/442

I hope circleci does its job this time.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader-base/1378)

### CODE REVIEW

File: cre.cpp [hunk 1]
```diff
@@ -3327,6 +3327,7 @@ static int findText(lua_State *L) {
 	int origin				= luaL_checkint(L, 3);
 	bool reverse			= luaL_checkint(L, 4);
 	bool caseInsensitive	= luaL_checkint(L, 5);
+	bool useRegex          	= luaL_checkint(L, 6);
```
Comment (poire-z) [inline cre.cpp:3330]: `isRegex`, `asRegex`, `patternIsRegex` ? `useRegex` feels odd.

File: cre.cpp [hunk 2]
```diff
@@ -3318,6 +3318,14 @@ static int regularizeRegisteredFontsWeights(lua_State *L) {
 	return 0;
 }
 
+static int checkRegex(lua_State *L) {
+	CreDocument *doc		= (CreDocument*) luaL_checkudata(L, 1, "credocument");
+	const char *l_pattern   = luaL_checkstring(L, 2);
+	lString32 pattern		= lString32(l_pattern);
+    lua_pushinteger(L, checkRegex(pattern));
```
Comment (poire-z) [inline cre.cpp:3325]: Some mixed indentations in there.
Do you think your checkRegex might return other values than 0 or 1? If no, better make this returns a boolean. If yes, well, ok :)

Comment (zwim) [reply cre.cpp:3325]: I would leave an int here. I would lik a more verbose error message here (we get that from the exception ).

File: Makefile.third [hunk 3]
```diff
@@ -204,6 +204,7 @@ $(CRENGINE_LIB): $(ZLIB) $(ZSTD_LIB) $(PNG_LIB) $(FREETYPE_LIB) $(HARFBUZZ_LIB)
 		-DUTF8PROC_INCLUDE_DIR="$(UTF8PROC_DIR)" \
 		-DZSTD_INCLUDE_DIR="$(ZSTD_DESTDIR)/include" \
 		-DCMAKE_BUILD_TYPE=$(CMAKE_BUILD_TYPE) \
+		-DSRELL_INCLUDE_DIR="${SRELL_INCLUDE_DIR}" \
```
Comment (poire-z) [inline Makefile.third:207]: Move it one line above, with the other *_INCLUDE_DIR.
-DCMAKE_BUILD_TYPE looks different :)

File: thirdparty/kpvcrlib/CMakeLists.txt [hunk 4]
```diff
@@ -47,6 +47,7 @@ include_directories(
     ${NANOSVG_INCLUDE_DIR}
     ${ZSTD_INCLUDE_DIR}
     ${CREGINE_INCLUDE_DIR}
+    ${SRELL_INCLUDE_DIR}
```
Comment (poire-z) [inline thirdparty/kpvcrlib/CMakeLists.txt:50]: Also move it above, the thirdparty _INCLUDE_DIR before the (mispelled) crengine own include dir - for consistency (dunno if it matters really :)

**Review by poire-z — COMMENTED** [review]
You bumped your crengine, and not your latest version. (No need to bump it until we merge it.)

**Comment by poire-z** [comment]
You added in thirdparty/srell the sources. Looks like it's not on GitHub. Do we have a specific way to handle this, or is it just fine to copy the source like done here? @Frenzie @NiLuJe ?

**Comment by zwim** [comment]
Another question about the license. Do we have to add the "License.txt" to our binary packages? Or is it sufficient that it is in the "thirdparty/srell/" folder?

**Comment by NiLuJe** [comment]
We usually take a shortcut and consider the source-code to be the "documentation" for the purpose of the BSD license. But it is admittedly a shortcut, and IANAL.

File: thirdparty/kpvcrlib/CMakeLists.txt [hunk 5]
```diff
@@ -71,6 +72,7 @@ add_definitions(
     -DALLOW_KERNING=1
     -DCR3_PATCH=1
     -DMATHML_SUPPORT=1
+    -DUSE_STD_REGEX=1
```
Comment (poire-z) [inline thirdparty/kpvcrlib/CMakeLists.txt:75]: This should now be called `-DUSE_SRELL` (and moved up with the other `-DUSE_*`).
Also update the `#if USE_STD_REGEX` in crengine.

File: Makefile.defs [hunk 6]
```diff
@@ -927,6 +927,11 @@ NANOSVG_DIR=$(CURDIR)/$(NANOSVG_BUILD_DIR)/nanosvg-prefix/src
 NANOSVG_INCLUDE_DIR=$(CURDIR)/$(NANOSVG_BUILD_DIR)/include
 NANOSVG_HEADERS=$(addprefix $(NANOSVG_INCLUDE_DIR)/, $(NANOSVG_HEADER_FILES))
 
+# Header files for
```
Comment (poire-z) [inline Makefile.defs:930]: `Srell headers (that include code)`

File: Makefile.defs [hunk 7]
```diff
@@ -927,6 +927,11 @@ NANOSVG_DIR=$(CURDIR)/$(NANOSVG_BUILD_DIR)/nanosvg-prefix/src
 NANOSVG_INCLUDE_DIR=$(CURDIR)/$(NANOSVG_BUILD_DIR)/include
 NANOSVG_HEADERS=$(addprefix $(NANOSVG_INCLUDE_DIR)/, $(NANOSVG_HEADER_FILES))
 
+# Srell headers (that include code)
+SRELL_DIR=$(CURDIR)/$(THIRDPARTY_DIR)/srell/
+SRELL_INCLUDE_DIR=$(SRELL_DIR)/
```
Comment (poire-z) [inline Makefile.defs:932]: Looks like most of the other *_DIR in this file don't have a trailing /.

File: Makefile.defs [hunk 8]
```diff
@@ -927,6 +927,11 @@ NANOSVG_DIR=$(CURDIR)/$(NANOSVG_BUILD_DIR)/nanosvg-prefix/src
...
+SRELL_HEADER_FILES=$(SRELL_INCLUDE_DIR)/srell.hpp
```
Comment (poire-z) [inline Makefile.defs:933]: Seems not used ?

File: cre.cpp [hunk 9]
```diff
@@ -3318,6 +3318,18 @@ static int regularizeRegisteredFontsWeights(lua_State *L) {
 	return 0;
 }
 
+static int checkRegex(lua_State *L) {
+	const char *l_pattern   = luaL_checkstring(L, 2);
+	lString32 pattern		= lString32(l_pattern);
```
Comment (poire-z) [inline cre.cpp:3323]: Strange/incosistent spacing (tabs?).
(It's bad in other places, but let's make the new code tab less, at least no tab in the middle :)

**Comment by zwim** [comment]
Sorry the last commit should be named bump crengine.

**Comment by poire-z** [comment]
Just to be sure and safe, with PRs that started a long time ago: do a last rebase to current master:
```
git checkout master
git pull
git checkout the_long_pr_branch
git rebase master
git push -f ...
```
And if this fails because your intermediate commits were walking all around, squash them to get a single commit
```
git checkout the_long_pr_branch
git rebase -i abcdef1234 (the hash before your first commit
   change all "pick" to "s" except the first
git log (you should see only one, with all the walking around commits having cancelled each other if needed)

and do the above rebase then
```

**Comment by zwim** [comment]
I have done something wrong
```Enter passphrase for key '/home/moscht/.ssh/id_rsa': 
To github.com:zwim/koreader-base.git
 ! [rejected]          findRegex -> findRegex (non-fast-forward)
error: Fehler beim Versenden einiger Referenzen nach 'github.com:zwim/koreader-base.git'
Hinweis: Aktualisierungen wurden zurückgewiesen, weil die Spitze Ihres aktuellen
Hinweis: Branches hinter seinem externen Gegenstück zurückgefallen ist. Führen Sie
Hinweis: die externen Änderungen zusammen (z. B. 'git pull ...') bevor Sie "push"
Hinweis: erneut ausführen.
Hinweis: Siehe auch die Sektion 'Note about fast-forwards' in 'git push --help'```

**Comment by zwim** [comment]
Is there something wrong? Now there is  https://github.com/koreader/koreader-base/pull/1378/commits/9710248b7c7b0268614b09a632d57a70fe5e329b in here?

**Comment by poire-z** [comment]
> I forgot the `-f` switch

No pb, you managed it the 2nd time.

But it still looks wrong: it looks like it cancels some previous commits, there is this `git merge upstream/master` that I don't want to see and you to use :)  (it was there too in some frontend PR of yours that messed things some weeks ago).
Try again my above suggestion just so you get confortable at sorting that kind of things - you can always open a new PR when you're fed up.

**Comment by poire-z** [comment]
Updating the first lines of my above commands:
```
git checkout master
git fetch upstream
git merge upstream/master
  so zwim master is iso with koreader master
git push origin master

git checkout the_long_pr_branch
git rebase master
git push -f ...
```

**Comment by zwim** [comment]
> Updating the first lines of my above commands:
> 
> ```
> git checkout master
> git fetch upstream
> git merge upstream/master
>   so zwim master is iso with koreader master
> git push origin master

If I do this and check zwim/koreader-base: I see ` This branch is 3 commits ahead of koreader:master. `
Is that ok?

> 
> git checkout the_long_pr_branch
> git rebase master
> git push -f ...
> ```

**Comment by zwim** [comment]
I will close this now and open a new PR.

**Comment by poire-z** [comment]
> If I do this and check zwim/koreader-base: I see This branch is 3 commits ahead of koreader:master.
Is that ok?

Yes, it was OK ! "ahead" is good, we just don't want to see any "behind".

Then, it's not just a matter of cleaning your 3 commits. (possibly dropping the one named "git merge upstream master" if it has nothing of yours.

Try it again if you have not deleted this branch.
Practice :) opening a new PR, you'll learn nothing and this will happen again.

**Comment by poire-z** [comment]
Well, may be not OK: your master should not be ahead. Why did this PR commits came into your master branch ?!
Anyway, you could have just dropped them - if they were also on your working branch.

**Pull request closed without merging**